irohad/ordering/impl/on_demand_ordering_service_impl.cpp

```diff
@@ -5,24 +5,104 @@
 
 #include "ordering/impl/on_demand_ordering_service_impl.hpp"
 
+#include "builders/protobuf/proposal.hpp"
+#include "interfaces/iroha_internal/proposal.hpp"
+#include "interfaces/transaction.hpp"
+
 using namespace iroha::ordering;
 
-// ------------------ | OnDemandOrderingService |-------------------------
+// -------------------------| OnDemandOrderingService |-------------------------
 
 void OnDemandOrderingServiceImpl::onCollaborationOutcome(
     RoundOutput outcome, transport::RoundType round) {
-  throw std::logic_error("Function not yet implemented");
+  // exclusive write lock
+  boost::upgrade_lock<boost::shared_mutex> upgrade_lock(lock_);
+  boost::upgrade_to_unique_lock<boost::shared_mutex> guard(upgrade_lock);
+
+  insertLast(outcome, round);
+  erase();
 }
 
-// -------------------- | OdOsNotification |------------------------------
+// ----------------------------| OdOsNotification |-----------------------------
 
 void OnDemandOrderingServiceImpl::onTransactions(
     std::vector<std::shared_ptr<shared_model::interface::Transaction>>
         transactions) {
-  throw std::logic_error("Function not yet implemented");
+  // read lock
+  boost::shared_lock<boost::shared_mutex> guard(lock_);
+
+  std::for_each(
+      transactions.begin(), transactions.end(), [this](const auto &obj) {
+        current_proposal_.second.push(obj);
+      });
 }
 
 boost::optional<std::shared_ptr<shared_model::interface::Proposal>>
-OnDemandOrderingServiceImpl::onRequestProposal(uint64_t round) {
-  throw std::logic_error("Function not yet implemented");
+OnDemandOrderingServiceImpl::onRequestProposal(transport::RoundType round) {
+  // read lock
+  boost::shared_lock<boost::shared_mutex> guard(lock_);
+  try {
+    return proposal_map_.at(round);
```

Comment:
since this section is protected by mutex, we can check if key is present and then request key. This avoids exception